python-package/lightgbm/dask.py

```diff
@@ -110,12 +110,128 @@ def _train_part(
     else:
         init_score = None
 
+    # construct local eval_set data.
+    local_eval_set = None
+    local_eval_names = None
+    local_eval_sample_weight = None
+    local_eval_class_weight = None
+    local_eval_init_score = None
+    local_eval_group = None
+    n_evals = max([len(x.get('eval_set', [])) for x in list_of_parts])
+    if n_evals:
+
+        has_eval_names = any([x.get('eval_names') is not None for x in list_of_parts])
+        has_eval_sample_weight = any([x.get('eval_sample_weight') is not None for x in list_of_parts])
+        has_eval_class_weight = any([x.get('eval_class_weight') is not None for x in list_of_parts])
+        has_eval_init_score = any([x.get('eval_init_score') is not None for x in list_of_parts])
+
+        local_eval_set = []
+        if has_eval_names:
+            local_eval_names = []
+        if has_eval_sample_weight:
+            local_eval_sample_weight = []
+        if has_eval_class_weight:
+            local_eval_class_weight = []
+        if has_eval_init_score:
+            local_eval_init_score = []
+        if is_ranker:
+            local_eval_group = []
+
+        # consolidate parts of each individual eval component.
+        for i in range(n_evals):
+            x_e = []
+            y_e = []
+            w_e = []
+            g_e = []
+            init_score_e = []
+            for part in list_of_parts:
+
+                if not part.get('eval_set'):
+                    continue
+
+                # possible that not each part contains parts of each individual (X, y) eval set.
+                if i >= len(part['eval_set']):
+                    continue
+
+                eval_set = part['eval_set'][i]
+                if eval_set == '__train__':
```

Comment:
I totally get what you're doing with these constants (e.g. `__train__`, `__sample_weight__`, etc.) and I support it. But each of these literal strings is defined twice in this module. Could you concentrate them in module-level constants that start with `_`?

Like

```python
_TRAIN_SET_PLACEHOLDER = '__train__'
```

Up to you if there's a more informative name than `_PLACEHOLDER` haha. I'm most concerned about keeping them consistent in both places. I also think that list of constants would be a good place to add a comment explaining something like "this is used by `lightgbm.dask`'s internals to say 'also evaluate the training data'. This is done to avoid duplicating the training data".

Reply:
Ha for sure! I like this idea a lot. Avatria use constants named like `_TRAIN_SET_NAME` a lot, so this makes sense, minimize number of lines that could require editing in the future.

Reply:
addressed in 2cb6973 - the `_DatasetNames` class might be controversial, just lmk

Reply:
looks good to me, thanks very much!